apps/src/pythonlab/patches/pythonScriptUtils.ts

```diff
@@ -54,59 +61,162 @@ export function writeSource(
+  ) as PyodidePathContent[];
+  const newSource = {...source};
+  updateAndDeleteSourceWithContents(
+    directoryContents,
+    newSource,
+    workingDir + '/',
+    DEFAULT_FOLDER_ID,
+    id,
+    pyodide,
+    sendMessage
+  );
+  return newSource;
+}
+
+function updateAndDeleteSourceWithContents(
+  contents: PyodidePathContent[],
+  source: MultiFileSource,
+  currentPath: string,
+  folderId: string,
+  id: string,
+  pyodide: PyodideInterface,
+  sendMessage: (message: PyodideMessage) => void
+) {
+  contents.forEach(content => {
+    const fileExtension = content.name.split('.').pop();
+    const fullPath = currentPath + content.name;
+    if (pyodide.FS.isFile(content.mode)) {
+      if (fileExtension === 'csv' || fileExtension === 'txt') {
+        const file = Object.values(source.files).find(
+          f => f.name === content.name
+        );
+        try {
+          const newContents = pyodide.FS.readFile(fullPath, {
+            encoding: 'utf8',
+          });
+          if (!file) {
+            const newFileId = getNextFileId(Object.values(source.files));
+            source.files[newFileId] = {
+              id: newFileId,
+              folderId,
+              name: content.name,
+              language: fileExtension,
+              contents: newContents,
+            };
+          } else {
+            file.contents = newContents;
+          }
+        } catch (e) {
+          console.warn(`could not read file ${content.name}, ${e}`);
+        }
+      } else {
+        if (!Object.values(source.files).some(f => f.name === content.name)) {
+          sendMessage({
+            type: 'error',
+            message: `You cannot write a ${fileExtension} file in Python Lab`,
+            id: id,
+          });
+        }
+      }
+      // Delete the file now that we have handled it.
+      pyodide.FS.unlink(fullPath);
+      // Do not create or iterate through folders that start with '.' as these are hidden folders.
+    } else if (
+      pyodide.FS.isDir(content.mode) &&
+      !content.name.startsWith('.')
+    ) {
+      const newPath = currentPath + content.name + '/';
+      const existingFolder = Object.values(source.folders).find(
+        f => f.name === content.name && f.parentId === folderId
+      );
+      let newFolderId = '';
+      if (!existingFolder) {
+        createFolderIfNotExists(newPath, pyodide);
+        newFolderId = getNextFolderId(Object.values(source.folders));
+        source.folders[newFolderId] = {
+          id: newFolderId,
+          name: content.name,
+          parentId: folderId,
+        };
+      } else {
+        newFolderId = existingFolder.id;
+      }
+
+      updateAndDeleteSourceWithContents(
+        Object.values(content.contents),
+        source,
+        newPath,
+        newFolderId,
+        id,
+        pyodide,
+        sendMessage
+      );
+      // Now that we've handled the contents, delete the folder.
+      pyodide.FS.rmdir(newPath);
```

Comment:
Do we need to add a follow-up task to handle errors on operations like `unlink` and `rmdir`?

Reply:
1. Good call I'll add to the follow-up task to handle errors here. 
2. I'm not sure. We delete everything in the working directory by the end of the function. I don't think there's a way for us to delete everything in bulk (`rmdir` only works on an empty directory, [api reference](https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.rmdir)). On my tests this process is very fast for a handful of files, and student projects probably aren't too big?

Reply:
Yeah, I think you'd need to add to "paths to delete" and then cycle through and rmdir or unlink them. I just had a neuron fire from a "deleting as you iterate" perspective. But it sounds like it's working fine and probably not an issue because we pass 
```
const directoryContents = Object.values(
    directoryData.contents
  ) as PyodidePathContent[];
  ``` 
  as an argument so it's not changing that array as we delete.

Reply:
yep exactly, I think there isn't a risk of the array changing from underneath us because we have a static object we are looking at.